x/concentrated-liquidity/pool_hooks.go

```diff
@@ -1,13 +1,118 @@
 package concentrated_liquidity
 
 import (
+	"encoding/json"
+
 	"github.com/cosmos/cosmos-sdk/store/prefix"
 	sdk "github.com/cosmos/cosmos-sdk/types"
 
+	"github.com/osmosis-labs/osmosis/osmomath"
+	"github.com/osmosis-labs/osmosis/osmoutils"
 	types "github.com/osmosis-labs/osmosis/v20/x/concentrated-liquidity/types"
 )
 
-// nolint: unused
+// Helper function to generate before action prefix
+func beforeActionPrefix(action string) string {
+	return "before" + action
+}
+
+// Helper function to generate after action prefix
+func afterActionPrefix(action string) string {
+	return "after" + action
+}
+
+// --- Pool Hooks ---
```

Comment:
Note: all of these pool hook functions are wrappers that build the appropriate message and then pass it into `callPoolActionListener` once marshaled. This structure abstracts 90%+ of the common logic out, and what is left is very difficult to generalize further without introducing complex helpers to build messages.

Thus, while some of this structure is repetitive, I believe this is an acceptable level of code complexity for introducing 10 hooks

Reply:
Can we instead reuse these helpers here somehow?

https://github.com/osmosis-labs/osmosis/blob/a0a4bd67d2a530addca628ddac29305fbe3838ae/osmoutils/cosmwasm/helpers.go#L113

I know it would require refactoring `callPoolActionListener` but I'm wondering whether it's worth it to reduce duplication

Reply:
Which duplication would this remove? The functions are just building and marshaling the custom message types (which would be required in any context) and all the other logic is abstracted by operating on marshaled message bytes

I might be missing your suggestion here though -- would you be able to elaborate further if the above misses the point?

Also, if the change is minor, it would probably be best to not overrotate on it since the downstream PR would need to be reimplemented as well

Reply:
`json.Marshal` is repetitive, and its duplication could be removed.

Agree though that this is not worth over-indexing on so already approved!